### PR TITLE
Create new cards for each filtered entity

### LINKED ISSUE #91: when using 'custom:slider-entitiy-row' as card in auto-entities, there is a Cannot read property 'split' of undefined
Hi,

First off all, all praise to the amazing custom cards you have done. I have been using HA for a while now and your cards are awesome.

As for this issue, i wanted to use auto-entities because it is cleaner/nicer then the entities filter and have a slider-entity-row show up if lights are on. I have tried just about every way i could think off but i keep getting an error:

`Cannot read property 'split' of undefined
entities:
  - entity: light.master_bedroom_spots
type: 'custom:slider-entity-row'`

Full code to configure it:

`  - type: "custom:auto-entities"
  card:
      type: "custom:slider-entity-row"
 filter:
   include:
      - entity_id: light.master_bedroom_spots
        state: "on"
  show_empty: false
`

I am more of a backend guy so maybe i missed something really stupid but 134 commits in. I am not seeing it :-s

### PR DESCRIPTION
**Pull request by Chropez**
The option `auto_cards` will create a new card for each filtered entity as an alternative to creating one card and passing multiple entities. The use case is to use the powerful filtering options of this component to dynamically create new cards for each entity.

resolves #91

### CODE REVIEW

**Comment by thomasloven** [comment]
This can be done if used in combination with https://github.com/thomasloven/lovelace-layout-card, which is actually preferable because that will make sure the cards are placed on screen in a   nicer way.

**Comment by Chropez** [comment]
I really like the api and power of filtering of this add-on and I wish it could create cards too. The reason being that there are so many add-ons that don't take an entities array as a prop, they just render a single entity. And unfortunately they all miss out of being able to be used in conjunction with this add-on. And it seems that that is a requested feature in the issues list.

For me it feels like the responsibility of filtering should lie in this add-on and leave up the layouting to other add-ons depending on their situation.

I've been using this addon new feature in a few different ways personally. Sometimes inside my own pure css responsive css-grid addon (I've seen examples where they have just used card-mod to achieve this) and some other times inside the addon `stack-in-card` to compose a card of multiple cards.
 
An example: I use this addon to filter my living room lights and render a `lovelace-slider-entity-row` for each of the lights. All those lights are then included in the `cards` prop of a `stack-in-card`. And for layouting the actual stack-in-cards.

```yaml
# See this as pseudo code
- custom:css-grid
  cards:
    - custom:stack-in-card nr1
       cards:
         - markdown-card
         - auto-enties filtering out all living room lights
            card:
              toggle: true
              type: custom:slider-entity-row
            auto_cards: true
            auto_cards_entity_param: entity

    - custom:stack-in-card nr2
```

Of course in the end, it's up to you if you believe that this feature aligns with your vision of this add-on. If not, I'll just keep using my forked repo ;)

Keep up with your awesome work!

**Comment by sgofferj** [comment]
Found this pull request through a Google search for auto creating cards. I'd second that function. I'm just doing some remodeling of my UI based on the new grid card and autocreating cards based on filters would be very helpful.

**Comment by ahochsteger** [comment]
@thomasloven I have the same requirement to create multiple cards (in my case <https://github.com/artem-sedykh/mini-climate-card>) where each card just take a single entity and wonder how `auto-entities` can be used in combination with `layout-card` as you suggested in the comment above?

**Comment by ahochsteger** [comment]
> @thomasloven I have the same requirement to create multiple cards (in my case https://github.com/artem-sedykh/mini-climate-card) where each card just take a single entity and wonder how `auto-entities` can be used in combination with `layout-card` as you suggested in the comment above?

Nevermind, just found it out by myself:
```yaml
type: 'custom:auto-entities'
filter:
  include:
    - domain: climate
      options:
        type: thermostat
  exclude:
    - state: unavailable
sort:
  method: name
  ignore_case: true
card:
  type: 'custom:layout-card'
  card_options:
    type: 'custom:mini-climate'
```

The tricky part was to find out that `card_options` has to be used to set the type of the cards that show a single entity. Would be great, if this could be added to the README as example.

**Comment by QNimbus** [comment]
> The tricky part was to find out that `card_options` has to be used to set the type of the cards that show a single entity. Would be great, if this could be added to the README as example.

Since the release of 2.0.0 on the 14th of March, the `card_options` parameter no longer seems to be available. 

Using this PR (#113 ) I managed to get my card (shown below) working. However using the other suggestion of using the `custom:layout-card` did not work because I was not able to pass `card_options` to the card. In fact, I did not manage to get the mediaplayer card but instead got the default `entities` card.

-- snippet --

```yaml
        - type: 'custom:auto-entities'
          show_empty: true
          auto_cards: false
          card_param: cards
          filter:
            entity_ids:
              - media_player.sonos_home_ff_office_mediaplayer_1
              - media_player.sonos_home_gf_living_mediaplayer_1
              - media_player.sonos_home_gf_mc_mediaplayer_1
              - media_player.sonos_home_ff_bedroom_mediaplayer_1
              - media_player.sonos_home_gf_mc_mediaplayer_2
            template: |
              {%- for mp in state_attr('media_player.sonos_home_ff_office_mediaplayer_1', 'sonos_group') | list -%}
                {%- if mp != 'media_player.sonos_home_ff_office_mediaplayer_1' -%}
                  {{- mp + ' ' -}}
                {%- endif -%}
              {%- endfor -%}
            type: 'custom:mini-media-player'
          card:
            type: 'custom:layout-card'
            layout_type: 'vertical-layout'
            card_options:
              type: 'custom:mini-media-player'
              group: true
              speaker_group:
                show_group_count: false
              hide:
                power: true
                source: true
                progress: true
                controls: true
                info: true
                icon: true
                volume_level: false
```

Below is the screenshot and the YAML of the configuration that _did_ work for me. This was using `auto-entities` with this pull request.

image

```yaml
      type: entities
      show_header_toggle: false
      entities:
        - type: 'custom:mini-media-player'
          entity: media_player.sonos_home_ff_office_mediaplayer_1
          hide:
            power: true
            volume_level: false
          info: scroll
          artwork: cover
          source: icon
          volume_step: 5
          background: /local/images/entity/sonos_amp.png
          group: true
          speaker_group:
            platform: sonos
            show_group_count: false
            sync_volume: false
            icon: 'hass:google-circles-communities'
            entities:
              - entity_id: media_player.sonos_home_ff_office_mediaplayer_1
                name: Werkkamer
              - entity_id: media_player.sonos_home_gf_living_mediaplayer_1
                name: Woonkamer
              - entity_id: media_player.sonos_home_gf_mc_mediaplayer_1
                name: Keuken
              - entity_id: media_player.sonos_home_ff_bedroom_mediaplayer_1
                name: Slaapkamer
              - entity_id: media_player.sonos_home_gf_mc_mediaplayer_2
                name: Badkamer
          tts:
            platform: sonos
        - type: 'custom:auto-entities'
          show_empty: true
          auto_cards: true
          filter:
            entity_ids:
              - media_player.sonos_home_ff_office_mediaplayer_1
              - media_player.sonos_home_gf_living_mediaplayer_1
              - media_player.sonos_home_gf_mc_mediaplayer_1
              - media_player.sonos_home_ff_bedroom_mediaplayer_1
              - media_player.sonos_home_gf_mc_mediaplayer_2
            template: |
              {%- for mp in state_attr('media_player.sonos_home_ff_office_mediaplayer_1', 'sonos_group') | list -%}
                {%- if mp != 'media_player.sonos_home_ff_office_mediaplayer_1' -%}
                  {{- mp + ' ' -}}
                {%- endif -%}
              {%- endfor -%}
          card:
            type: 'custom:mini-media-player'
            group: true
            speaker_group:
              show_group_count: false
            hide:
              power: true
              source: true
              progress: true
              controls: true
              info: true
              icon: true
              volume_level: false
```

TL;DR I think that this feature is really usefull for when you want to use the `auto-enties` card with multiple cards.

**Comment by adabelleleiram** [comment]
Any updates on this? Either the solution of combining getting it to work with the layout-card or if this PR will be merged?

I'm using this card to get a template list of entities and would like to put them each in a customised button card but don't seem to be able to solve it currently (especially since  card_options doesn't exist in the layout-card)
@thomasloven

**Comment by thomasloven** [comment]
https://github.com/thomasloven/lovelace-auto-entities/wiki/Template-filters

**Comment by adabelleleiram** [comment]
> https://github.com/thomasloven/lovelace-auto-entities/wiki/Template-filters

So you can only set the type of the card? Not add any e.g. styling?

**Comment by thomasloven** [comment]
What makes you think it would suddenly stop working?

**Comment by adabelleleiram** [comment]
Oh sorry, I copied the second example to my config and forgot there was another one. I blame toddler-induced sleep deprivation 🙈 

Last question, is there any way of getting all the created cards in a horizontal stack? With the layout card, they all get split up.

**Comment by thomasloven** [comment]
>  toddler-induced sleep deprivation

I know that feeling.

[layout-card](https://github.com/thomasloven/lovelace-layout-card) can place cards in any layout.
